README.md
# Project Switch - Server

__Server__ is the backend for the __Project Switch__. The server publishes RESTful APIs which are written using __flask__ and communication with the hardware is done using __MQTT__.

## Getting Started

This server is currently deployed on __raspberrypi 3__. It can also be deployed on any Debian based Linux distribution.

- Clone this repo on to your local machine

```bash
git clone git@gitlab.com:project-switch/server.git
```

### Environment

Things you need to deploy this application

- Raspbian or Debian GNU/Linux

- Python 3.7

- Mosquitto Broker

- Virtualenv

### Installing

#### MQTT Broker

- Install the Mosquitto Broker

```bash
sudo apt update
sudo apt install -y mosquitto mosquitto-clients
```

- Make Mosquitto auto start on boot up enter

```bash
sudo systemctl enable mosquitto.service
```

#### Database

- Install SQLite3 db

```bash
sudo apt install sqlite3
```

#### Python Environment

- Go to the project root directory and create a Virtual environment to run this application as `switch_env`

```bash
cd server
python3 -m venv switch_env
```

- Activate the python environment

```bash
source switch_env/bin/activate
```

- Install python packages for the backend

```bash
pip3 install -r requirements.txt
```

- Now run the server

```bash
python3 wsgi.py
```

The server will be running on `http://localhost:5000/`

- To deactivate the environment just type

```bash
deactivate
```

## Deployment

- Run this Application using __gunicon__

```bash
gunicorn --bind 0.0.0.0:5000 wsgi:app
```

Now create a service file to run this application on restart

## Contributing

Please read [CONTRIBUTING](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Changelog

Check [CHANGELOG](CHANGELOG.md) to get the version details.

## License

This project is licensed under the GNU AGPLv3 License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# Changelog

We use  for versioning.
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). It diverges in the following ways:

- Release titles do not link to the commits within the release
- This project only strictly adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) for bug fix releases.

## [1.0.0] - 2020-10-03

### Added

- RESTful APIs

- MQTT communication with Hardware

- User Authentication


smartHome/controllers/users.py
from smartHome import db_location
from flask import Blueprint, jsonify, request
import sqlite3


user_route = Blueprint('user', __name__)


def find_user_by_username(username):
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "SELECT * FROM users WHERE username=?"
    result = cursor.execute(query, (username,))
    row = result.fetchone()
    connection.close()
    if row:
        return row


def find_topics_by_room(room):
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "SELECT * FROM topics WHERE room=?"
    row = cursor.execute(query, (room,))
    if row:
        keys = ['room', 'device', 'topic', 'status', 'icon']
        deviceData = [dict(zip(keys, row)) for row in cursor.fetchall()]
        connection.close()
        return deviceData


def find_all_topics():
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "SELECT * FROM topics"
    row = cursor.execute(query, ())
    if row:
        keys = ['room', 'device', 'topic', 'status', 'icon']
        deviceData = [dict(zip(keys, row)) for row in cursor.fetchall()]
        connection.close()
        return deviceData


@user_route.route('/api/test')
def testRoute():
    return jsonify({
        "status": 200,
        "message": "You are in Test Route"
    })


@user_route.route('/api/auth', methods=["POST"])
def getDevices():
    requestData = request.get_json()
    user = requestData['username']
    password = requestData['password']
    try:
        dbUser = find_user_by_username(user)
        if dbUser[1] == password:
            if dbUser[0] == 'admin':
                deviceData = find_all_topics()
                data = {"room": dbUser[2], "devices": deviceData}
                pass
            else:
                deviceData = find_topics_by_room(dbUser[2])
                data = {"room": dbUser[2], "devices": deviceData}
            return jsonify({
                "message": "You are Authenticated",
                "data": data
            }), 200
        else:
            return jsonify({
                "message": "You are Not Authenticated"
            }), 401
    except Exception as e:
        print("Oops!", e.__class__, "occurred.")
        return jsonify({
            "message": "You are Not Authenticated"
        }), 500


smartHome/controllers/mqttClient.py
from smartHome import mqttc, db_location
from flask import Blueprint, jsonify, request
import sqlite3


mqtt_route = Blueprint('mqtt_route', __name__)


def find_topics_by_room(room):
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "SELECT * FROM topics WHERE room=?"
    row = cursor.execute(query, (room,))
    if row:
        keys = ['room', 'device', 'topic', 'status', 'icon']
        deviceData = [dict(zip(keys, row)) for row in cursor.fetchall()]
        connection.close()
        return deviceData


def find_topic_by_room_and_device(room, device):
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "SELECT * FROM topics WHERE room=? AND device=?"
    result = cursor.execute(query, (room, device))
    row = result.fetchone()
    connection.close()
    if row:
        return row


def find_device_and_update(action, room, device):
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "UPDATE topics SET status=? WHERE room=? AND device=?"
    cursor.execute(query, (action, room, device))
    connection.commit()
    connection.close()


def update_all(action, room):
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "UPDATE topics SET status=? WHERE room=?"
    cursor.execute(query, (action, room))
    connection.commit()
    connection.close()


def publishData(requestData):
      room = requestData['room']
      device = requestData['device']
      action = requestData['action']
      try:
          dbTopic = find_topic_by_room_and_device(room, device)
          if dbTopic:
            if action == "ON":
                mqttc.publish(dbTopic[2], "0")
                print('Published : {}'.format(dbTopic))
            if action == "OFF":
                mqttc.publish(dbTopic[2], "1")
                print('Published : {}'.format(dbTopic))
            find_device_and_update(action, room, device)
            deviceData = find_topics_by_room(room)
            data = {"room": room, "devices": deviceData}
            return {
                "message": "Request Sucessful",
                "data": data
            }, 200
          else:
            return {"message": "Server Error"}, 500
      except Exception as e:
         print("Oops!", e.__class__, "occurred.")
         return {"message": "Server Error"}, 500


def statusRest(requestData):
    room = requestData['room']
    status = requestData['reset']
    try:
      if status == "ON":
        action = "0"
      elif status == "OFF":
        action = "1"
      devices = find_topics_by_room(room)
      for device in devices:
        mqttc.publish(device['topic'], action)
      update_all(status, room)
      print("Room Reset")
      return { "message": "Request Sucessful" }, 200
    except Exception as e:
        print("Oops!", e.__class__, "occurred.")
        return {"message": "Server Error"}, 500


@mqtt_route.route("/api/dashboard", methods=['POST'])
def dashboard():
   requestData = request.get_json()
   if requestData:
      message, statusCode = publishData(requestData)
      return jsonify(message), statusCode
   else:
      return jsonify({
         "message": "Bad Request"
      }), 400


@mqtt_route.route("/api/dashboard/reset", methods=['POST'])
def resetAll():
   requestData = request.get_json()
   if requestData:
      message, statusCode = statusRest(requestData)
      return jsonify(message), statusCode
   else:
      return jsonify({
         "message": "Bad Request"
      }), 400


smartHome/__init__.py
from flask import Flask, request, jsonify
from flask_cors import CORS, cross_origin
import paho.mqtt.client as mqtt
import sqlite3
import os


app = Flask(__name__)
cors = CORS(app, resources={r"/api/*": {"origins": "*"}})
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
db_location = BASE_DIR + '/smartHome.db'

mqttc = mqtt.Client()
mqttc.connect("localhost", 1883, 60)
mqttc.loop_start()


# Blueprints
from smartHome.controllers.mqttClient import mqtt_route
from smartHome.controllers.users import user_route


app.register_blueprint(user_route)
app.register_blueprint(mqtt_route)


def find_all_topics():
    connection = sqlite3.connect(db_location)
    cursor = connection.cursor()
    query = "SELECT * FROM topics"
    row = cursor.execute(query)
    if row:
        keys = ['room', 'device', 'topic', 'status', 'icon']
        devices = [dict(zip(keys, row)) for row in cursor.fetchall()]
        connection.close()
        return devices


def action_on_reboot():
    devices = find_all_topics()
    for device in devices:
        if device['status'] == "ON":
            action = "0"
        elif device['status'] == "OFF":
            action = "1"
        mqttc.publish(device['topic'], action)
        print('Published {}'.format(device))


action_on_reboot()


@app.route("/")
def main():
    return jsonify({
        "messages": "Sever Running on Port 5000"
    }), 200


@app.route("/robots.txt")
def robots_dot_txt():
    return "User-agent: *\nDisallow: /"
